kafka-rest/src/main/java/io/confluent/kafkarest/resources/RateLimiter.java

```diff
@@ -35,7 +34,6 @@
 
   private final int maxRequestsPerSecond;
   private final long gracePeriod;
-  private final boolean rateLimitingEnabled;
   private final Clock clock;
   private final AtomicInteger rateCounterSize = new AtomicInteger(0);
```

Comment:
Each rate limiter has a non-trivial memory footprint dominated by the deque of boxed longs. Do you have an estimate for the upper bound of the number of logical cluster in an MT cluster? Can we figure out what might be the max impact this might have on the MT cluster memory? Back of the envelope with totally imaginary numbers, 10k longs per limiter, 1k limiter per MT cluster mean ~80 MB. That feels realistic and not too concerning, I just want to make sure I'm not overlooking an order of magnitude somewhere.

Each rate limiter will also be doing a good amount of enqueuing and dequeuing from its deque. We'll be seeing what would be the effect of that on CPU when we actually do some experiments. If CPU suffers while Memory seems to be OK, we can use that trade-off to some extent - implement an array-based (preallocated) queue which allows binary-search drain, or use binning for a histogram-like rate limit.
- The histogram-like rate limit makes sense anyway if we are OK with applying/updating our rate limits every X milliseconds, which sounds pretty acceptable to me. If it's HDR, it can also address our previous memorylessness concern.

Reply:
See https://confluent.slack.com/archives/C3226NBLJ/p1635953342157700?thread_ts=1635951301.157600&cid=C3226NBLJ (as this is a public git repo), for actual numbers of clusters and I agree with your calculations.

Wrt to the rate limiting alogrithm itself, it's implementation not externals, so we can revisit it in the future. Again, lets chat some more on Monday but this shouldn't block this PR I hope?

Reply:
Chatting on Monday SGTM.